collector/zfs.go

```diff
@@ -0,0 +1,179 @@
+package collector
+
+// +build !nofilesystem
+
+import (
+	"fmt"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+type zfsSysctl string
+type zfsComputedMetric string
+type zfsMetricValue int
+
+const zfsErrorValue = zfsMetricValue(-1)
+
+type zfsMetricProvider struct {
+	fetchedResults  map[zfsSysctl]zfsMetricValue
+	computedResults map[zfsComputedMetric]zfsMetricValue
+}
+
+type zfsCollector struct {
+	metricProvider  *zfsMetricProvider
+	fetchedMetrics  map[zfsSysctl]prometheus.Gauge
+	computedMetrics map[zfsComputedMetric]prometheus.Gauge
+}
+
+//------------------------------------------------------------------------------
+//                                 Collector
+//------------------------------------------------------------------------------
+
+func init() {
+	Factories["zfs"] = NewZFSCollector
+}
+
+const zfsArcSubsystemName = "zfs_arc"
+
+func NewZFSCollector() (Collector, error) {
+	if err := zfsInitialize(); err != nil {
+		return &zfsCollector{}, err
+	}
+
+	fetchedMetrics := make(map[zfsSysctl]prometheus.Gauge)
+
+	insertGauge := func(sysctl, name string) {
+		fetchedMetrics[zfsSysctl(sysctl)] = prometheus.NewGauge(
```

Comment:
You should be using ConstMetrics rather than creating Gauges. Gauges are the like are for direction instrumentation, whereas a custom collector like this should use ConstMetrics.